{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# h1: first haplotype\n",
    "# h2: second haplotype\n",
    "# s: effect size of variants\n",
    "gene_expression_estimation<-function(h1,h2,s){\n",
    "  log_expression_h1<-t(h1)%*%s\n",
    "  log_expression_h2<-t(h2)%*%s\n",
    "  total_expression<-exp(log_expression_h1)+exp(log_expression_h2)\n",
    "  expression_ratio_h1<-exp(log_expression_h1)/total_expression\n",
    "  expression_ratio_h2<-exp(log_expression_h2)/total_expression\n",
    "  result<-c(log_expression_h1,log_expression_h2,total_expression,expression_ratio_h1,expression_ratio_h2)\n",
    "  return(result)\n",
    "  \n",
    "}\n",
    "\n",
    "# producing all genotypes\n",
    "number2binary = function(number, noBits) {\n",
    "       binary_vector = rev(as.numeric(intToBits(number)))\n",
    "       if(missing(noBits)) {\n",
    "          return(binary_vector)\n",
    "       } else {\n",
    "          binary_vector[-(1:(length(binary_vector) - noBits))]\n",
    "       }\n",
    "    }\n",
    "\n",
    "#lookup table for n variants\n",
    "\n",
    "AFC=read.table(file=\"aFC_Whole_Blood.txt\", header=TRUE, sep=\"\\t\")\n",
    "# maximum number of variants\n",
    "maximum_variants <- 16\n",
    "\n",
    "for (var_count in 1:maximum_variants){    \n",
    "    lookup_table<-data.frame()\n",
    "    lookup_table_index<-1\n",
    "    count<-1\n",
    "    result<-c()\n",
    "    count_result<-c()\n",
    "    for (index in 1:nrow(AFC)){\n",
    "        \n",
    "        # check the number of variants \n",
    "        current_gene<-paste(AFC[index,'gene_id'])\n",
    "        print (current_gene)\n",
    "        next_gene<-paste(AFC[index+1,'gene_id'])\n",
    "        \n",
    "        if(identical(current_gene,next_gene)){\n",
    "            count<-count+1\n",
    "            \n",
    "            }\n",
    "        else{\n",
    "            \n",
    "            if (var_count==count){\n",
    "                    \n",
    "                gene_id<-gsub(\"\\\\..*\",\"\",AFC[index,'gene_id'])\n",
    "                lookup_table[lookup_table_index,'gene_id']<-paste0(gene_id)\n",
    "                AFC_vector<-c()\n",
    "                var_index<-1\n",
    "                    for (temp_index in (index-count+1):index){\n",
    "                        AFC_value<-AFC[temp_index,'log2_aFC']\n",
    "                        \n",
    "                        \n",
    "                        \n",
    "                        if (AFC_value=='NaN'| AFC_value=='nan' | is.na(AFC_value))\n",
    "                            AFC_value<-0\n",
    "                        \n",
    "                        \n",
    "                        # Cut off afc values\n",
    "                        if (AFC_value < -6.64){\n",
    "                            AFC_value <- -log2(100)\n",
    "                            }\n",
    "                        if (AFC_value > 6.64){\n",
    "                            AFC_value <- log2(100)\n",
    "                            }\n",
    "                        # Cut off afc values\n",
    "                        \n",
    "                        AFC_vector<- c(AFC_vector,AFC_value)\n",
    "                        \n",
    "                        # create lookup table\n",
    "                        \n",
    "                        lookup_table[lookup_table_index,paste0('variant_id_',var_index)]<-paste0(AFC[temp_index,'variant_id'])\n",
    "\n",
    "                        lookup_table[lookup_table_index,paste0('rank_',var_index)]<-paste0(AFC[temp_index,'rank'])\n",
    "                        var_index<-var_index+1    \n",
    "                        }\n",
    "                    for (j in 0:(2^count-1)){\n",
    "                        \n",
    "                        ref_alt_vector<-c(number2binary(j,count))\n",
    "                        ref_vector<-c(replicate(count,0))\n",
    "                        output<-gene_expression_estimation(ref_alt_vector,ref_vector,AFC_vector)\n",
    "                        lookup_table[lookup_table_index,paste0(\"\",toString(ref_alt_vector))]<-paste0(output[1])\n",
    "                        \n",
    "                        \n",
    "                        \n",
    "                    }\n",
    "                   \n",
    "                    lookup_table_index<-lookup_table_index+1   \n",
    "                \n",
    "                \n",
    "                \n",
    "                }\n",
    "            count<-1\n",
    "            }#end of else\n",
    "        }\n",
    "# create output  \n",
    "    \n",
    "    write.csv(lookup_table, file=paste0(\"output\",var_count,\".csv\"))\n",
    "    write.table(lookup_table,file=paste0(\"output\",var_count,\".txt\"))\n",
    "    \n",
    "    \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": false,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
