{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import islice\n",
    "import os\n",
    "\n",
    "def genotypeOfHaplotypes(genotype_info):\n",
    "    \n",
    "    FirstPart = genotype_info.split(\":\")[0];\n",
    "    \n",
    "    if not (\"|\" in FirstPart):\n",
    "        return \"22\";\n",
    "    \n",
    "    FirstPartSplit = FirstPart.split(\"|\");\n",
    "    \n",
    "    genotype_hap1 = 0 if FirstPartSplit[0] == \"0\" else 1;\n",
    "    genotype_hap2 = 0 if FirstPartSplit[1] == \"0\" else 1;\n",
    "    return (str(genotype_hap1) + str(genotype_hap2));\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "maximum_number_varinats = 16\n",
    "# where individuals id start in vcf file\n",
    "individual_starting_column = 12 \n",
    "for variationCount in range(1,maximum_number_varinats):\n",
    "\n",
    "\n",
    "\n",
    "    lookup_table= pd.read_table(\"lookupTable_variantNo_\"+str(variationCount)+\".txt\", sep=\" \");\n",
    "    genotype_info= pd.read_table(\"data\\vcf_sample.txt\", sep = \"\\t\")\n",
    "# sort lookup table \n",
    "    lookup_table = lookup_table.sort_values(by=['variant_id_1'])    \n",
    "\n",
    "    lookup_table = lookup_table.reset_index(drop= True)  #make indexes start at 0\n",
    "\n",
    "    \n",
    "    prediction_matrix_haplotype1= pd.DataFrame()\n",
    "    prediction_matrix_haplotype2= pd.DataFrame()\n",
    "\n",
    "    hap1_individual_genotype_lst = []\n",
    "    hap2_individual_genotype_lst = []\n",
    "    temp = pd.DataFrame();\n",
    "    \n",
    "    pred_index =1 \n",
    "    individuals = genotype_info.columns.tolist()[individual_starting_column:]\n",
    "    individual_info_hap1 = pd.DataFrame(columns = individuals)\n",
    "    individual_info_hap2 = pd.DataFrame(columns = individuals)\n",
    "\n",
    "    genotype_info_hap1 = pd.DataFrame(columns = individuals)\n",
    "    genotype_info_hap2 = pd.DataFrame(columns = individuals)\n",
    "\n",
    "    for index, row in lookup_table.iterrows():\n",
    "    \n",
    "        hap1_individual_genotype_lst.clear()\n",
    "        hap2_individual_genotype_lst.clear()\n",
    "        columns = pd.Series()\n",
    "        columns = columns.append(pd.Series({'gene_id':row['gene_id']}))\n",
    "        for i in range(1,variationCount+1):\n",
    "            var= 'variant_id_' + str(i)\n",
    "            columns = columns.append(pd.Series({var:row[var]}))\n",
    "            temp = columns.to_frame().T \n",
    "            key = row['gene_id']+ \"_\"+row[var]\n",
    "            print (key)\n",
    "        \n",
    "# find the gene in genotype file\n",
    "        \n",
    "            genotype_index = list(np.where(genotype_info[\"gene_id_ID\"] == key))[0]\n",
    "        \n",
    "            if (len(genotype_index)==0):     # chrX won't be found\n",
    "                continue\n",
    "#print (genotype_index[0]);\n",
    "            hap_ind = 0\n",
    "        \n",
    "            for ind in individuals:\n",
    "            \n",
    "# make genotypes            \n",
    "                genotype_both_haps= genotypeOfHaplotypes(genotype_info.loc[genotype_index[0],ind])\n",
    "            \n",
    "                if (i>1):\n",
    "                    hap1_individual_genotype_lst.insert(hap_ind, hap1_individual_genotype_lst[hap_ind]+str(genotype_both_haps[0]) + \", \")\n",
    "                    del hap1_individual_genotype_lst[hap_ind + 1]\n",
    "                    hap2_individual_genotype_lst.insert(hap_ind, hap2_individual_genotype_lst[hap_ind]+str(genotype_both_haps[1]) + \", \")\n",
    "                    del hap2_individual_genotype_lst[hap_ind + 1]\n",
    "                else:\n",
    "                    hap1_individual_genotype_lst.insert(hap_ind, str(genotype_both_haps[0]) + \", \")\n",
    "                    hap2_individual_genotype_lst.insert(hap_ind, str(genotype_both_haps[1]) + \", \")\n",
    "                hap_ind = hap_ind+1\n",
    "   \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "        prediction_matrix_haplotype1= prediction_matrix_haplotype1.append(temp,ignore_index=True)\n",
    "        prediction_matrix_haplotype2= prediction_matrix_haplotype2.append(temp,ignore_index=True)\n",
    "    \n",
    "        i = 0\n",
    "        for genotype in hap1_individual_genotype_lst:\n",
    "        \n",
    "            genotype_info_hap1.at[index, individuals[i]] =  genotype[:-2]\n",
    "            if (\"2\" in genotype[:-2] ):\n",
    "# non-phased data\n",
    "                individual_info_hap1.at[index, individuals[i]] = \"NA\"\n",
    "        \n",
    "            else:\n",
    "# find the genotype in lookup table\n",
    "                individual_info_hap1.at[index, individuals[i]] =  row[genotype[:-2]]\n",
    "            i = i+1\n",
    "    \n",
    "        i=0;    \n",
    "        for genotype in hap2_individual_genotype_lst:\n",
    "        \n",
    "            genotype_info_hap2.at[index, individuals[i]] =  genotype[:-2]\n",
    "            if (\"2\" in genotype[:-2] ):\n",
    "                individual_info_hap2.at[index, individuals[i]] = \"NA\"\n",
    "        \n",
    "            else:\n",
    "        \n",
    "                individual_info_hap2.at[index, individuals[i]] =  row[genotype[:-2]]\n",
    "            i = i+1\n",
    "        \n",
    "        \n",
    "        \n",
    "    output_hap1 = pd.concat([prediction_matrix_haplotype1, individual_info_hap1],axis = 1, sort = False)\n",
    "    output_hap2 = pd.concat([prediction_matrix_haplotype2, individual_info_hap2],axis = 1, sort = False)\n",
    "\n",
    "    output_genotype_hap1 = pd.concat([prediction_matrix_haplotype1, genotype_info_hap1],axis = 1, sort = False)\n",
    "    output_genotype_hap2 = pd.concat([prediction_matrix_haplotype2, genotype_info_hap2],axis = 1, sort = False)\n",
    "\n",
    "#outputs\n",
    "    output_hap1.to_csv('ASE_prediction_hap1_Variant_Num_'+str(variationCount)+\".csv\")\n",
    "    output_hap2.to_csv('ASE_prediction_hap2_Variant_Num_'+str(variationCount)+\".csv\") \n",
    "    output_genotype_hap1.to_csv('genotype_hap1_Variant_Num_'+str(variationCount)+\".csv\")\n",
    "    output_genotype_hap2.to_csv('genotype_hap2_Variant_Num_'+str(variationCount)+\".csv\")    \n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": false,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
